criu/cr-dump.c

```diff
@@ -1183,107 +1256,26 @@ static int dump_task_rseq(pid_t pid, struct pstree_item *item)
-	if (!rseq_cs->start_ip)
-		return 0;
-
-	pr_debug(
-		"fixup_thread_rseq for %d: rseq_cs start_ip = %llx abort_ip = %llx post_commit_offset = %llx flags = %x version = %x; IP = %lx\n",
-		tid, rseq_cs->start_ip, rseq_cs->abort_ip, rseq_cs->post_commit_offset, rseq_cs->flags,
-		rseq_cs->version, (unsigned long)TI_IP(core));
-
-	if (rseq_cs->version != 0) {
-		pr_err("unsupported RSEQ ABI version = %d\n", rseq_cs->version);
-		return -1;
-	}
-
-	if (task_in_rseq(rseq_cs, TI_IP(core))) {
-		struct pid *tid = &item->threads[i];
-
-		/*
-		 * We need to fixup task instruction pointer from
-		 * the original one (which lays inside rseq critical section)
-		 * to rseq abort handler address. But we need to look on rseq_cs->flags
-		 * (please refer to struct rseq -> flags field description).
-		 * Naive idea of flags support may be like... let's change instruction pointer (IP)
-		 * to rseq_cs->abort_ip if !(rseq_cs->flags & RSEQ_CS_FLAG_NO_RESTART_ON_SIGNAL).
-		 * But unfortunately, it doesn't work properly, because the kernel does
-		 * clean up of rseq_cs field in the struct rseq (modifies userspace memory).
-		 * So, we need to preserve original value of (struct rseq)->rseq_cs field in the
-		 * image and restore it's value before releasing threads (see restore_rseq_cs()).
-		 *
-		 * It's worth to mention that we need to fixup IP in CoreEntry
-		 * (used when full dump/restore is performed) and also in
-		 * the parasite regs storage (used if --leave-running option is used,
-		 * or if dump error occurred and process execution is resumed).
-		 */
-
-		if (!(rseq_cs->flags & RSEQ_CS_FLAG_NO_RESTART_ON_SIGNAL)) {
-			pr_warn("The %d task is in rseq critical section. IP will be set to rseq abort handler addr\n",
-				tid->real);
-
-			TI_IP(core) = rseq_cs->abort_ip;
-
-			if (item->pid->real == tid->real) {
-				compel_set_leader_ip(dmpi(item)->parasite_ctl, rseq_cs->abort_ip);
-			} else {
-				compel_set_thread_ip(dmpi(item)->thread_ctls[i], rseq_cs->abort_ip);
-			}
-		}
-	}
-
-	return 0;
-}
-
-static int fixup_task_rseq(pid_t pid, struct pstree_item *item)
-{
-	int ret = 0;
-	int i;
-
-	if (!kdat.has_ptrace_get_rseq_conf)
-		return 0;
-
-	for (i = 0; i < item->nr_threads; i++) {
-		if (fixup_thread_rseq(item, i)) {
-			ret = -1;
-			goto exit;
-		}
-	}
-
-exit:
-	xfree(dmpi(item)->thread_rseq_cs);
-	dmpi(item)->thread_rseq_cs = NULL;
-	return ret;
-}
-
 static struct proc_pid_stat pps_buf;
 
 static int dump_task_threads(struct parasite_ctl *parasite_ctl, const struct pstree_item *item)
 {
-	int i;
+	int i, ret = 0;
 
 	for (i = 0; i < item->nr_threads; i++) {
 		/* Leader is already dumped */
 		if (item->pid->real == item->threads[i].real) {
 			item->threads[i].ns[0].virt = vpid(item);
 			continue;
 		}
-		if (dump_task_thread(parasite_ctl, item, i))
-			return -1;
+		ret = dump_task_thread(parasite_ctl, item, i);
+		if (ret)
+			break;
 	}
 
-	return 0;
+	xfree(dmpi(item)->thread_rseq_cs);
+	dmpi(item)->thread_rseq_cs = NULL;
```

Comment:
Not related to your change. But in fact there is a memory leak. And that's my bad in this case.

We are allocating `dmpi(item)->thread_rseq_cs` in `dump_task_rseq`. Then:
```

	ret = dump_task_rseq(pid, item); // <<-- allocated
	if (ret) {
		pr_err("Dump %d rseq failed %d\n", pid, ret);
		goto err;
	}

	parasite_ctl = parasite_infect_seized(pid, item, &vmas);
	if (!parasite_ctl) {
		pr_err("Can't infect (pid: %d) with parasite\n", pid);
		goto err;
	}

//... a lot of code...

	ret = dump_task_threads(parasite_ctl, item); // <-- freed
	if (ret) {
		pr_err("Can't dump threads\n");
		goto err_cure;
	}
```

If anything goes wrong in between `dump_task_rseq` and `dump_task_threads` then we have a memleak. And I think it's not only about `dmpi(item)->thread_rseq_cs` but also about `dmpi(item)->thread_ctls` and `dmpi(item)->thread_sp` too...

That's ToDo for us :-)